### PR TITLE
Check entity actions in submission against those permitted by form def

### PR DESCRIPTION
**Pull request by matthew-white**
Closes getodk/central#518.

#### What has been done to verify that this works as intended?

New tests.

#### Why is this the best possible solution? Were any other approaches considered?

The permitted actions from the form def are persisted on the `dataset_form_defs` record. It seems like a good idea to persist the actions rather than parsing the form XML each time a submission is processed for entities. `dataset_form_defs` also seems like the right table to store the actions. In the future, a form def will be able to take actions on multiple datasets, and in theory, it could take different actions on the different datasets. So conceptually, the permitted actions aren't per form def: they're per form def + dataset.

Note also that a single form may have different versions/defs that permit different actions. When a submission is processed, the form version specified in the submission will be used.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

All existing forms that make changes to entities will permit entity creation and only entity creation.

One edge case that was tolerated before now results in an error. See test/unit/data/dataset.js for details about that change. I think it's likely good that we no longer tolerate that edge case.

#### Does this change require updates to the API documentation? If so, please update docs/api.md as part of this PR.

No API responses have changed (other than some Problems). Maybe one day we'll want to return the list of permitted actions over the API.

#### Before submitting this PR, please make sure you have:

- [x] run `make test-full` and confirmed all checks still pass OR confirm CircleCI build passes
- [x] verified that any code from external sources are properly credited in comments or that everything is internally sourced

### CODE REVIEW

File: lib/util/problem.js [hunk 1]
```diff
@@ -136,6 +136,9 @@ const problems = {
     // TODO: should have details but not sure what yet.
     insufficientRights: problem(403.1, () => 'The authentication you provided does not have rights to perform that action.'),
 
+    entityActionNotAllowed: problem(403.2, ({ action, permitted }) =>
+      `The submission attempts an entity ${action}, but the form does not permit that action. The form permits the following actions: ${permitted.join(',')}.`),
```
Comment (matthew-white) [inline lib/util/problem.js:140]: `', '`

File: lib/data/dataset.js
Comment (matthew-white) [inline lib/data/dataset.js:1]: Update the comments above `getDataset()`.

File: lib/data/dataset.js [hunk 2]
```diff
@@ -61,21 +61,36 @@ const getDataset = (xml) => {
   const metaNode = findOne(root('html'), node('head'), node('model'), node('instance'), node(), node('meta'));
   return traverseXml(xml, [
     findOne(root('html'), node('head'), node('model'))(attr('entities-version')),
-    metaNode(findOne(root(), node('entity'))(tree())),
-    metaNode(findOne(root(), and(node('entity'), hasAttr('dataset')))(attr('dataset')))
-  ]).then(([ version, entityTag, datasetName ]) => {
-    if (entityTag.isEmpty() && datasetName.isEmpty())
+    metaNode(findOne(root(), node('entity'))(attr()))
```
Comment (matthew-white) [inline lib/data/dataset.js:64]: Move the `metaNode` here since it's no longer referenced twice.

File: lib/data/dataset.js [hunk 3]
```diff
@@ -61,21 +61,36 @@ const getDataset = (xml) => {
...
     else if (!(version.get().startsWith('2022.1.') || version.get().startsWith('2023.1.')))
       throw Problem.user.invalidEntityForm({ reason: `Entities specification version [${version.get()}] is not supported.` });
+
+    const strippedAttrs = Object.create(null);
+    for (const [name, value] of Object.entries(entityAttrs.get()))
+      strippedAttrs[stripNamespacesFromPath(name)] = value;
+
     // check that dataset name is valid
-    if (datasetName.isEmpty())
+    const datasetName = strippedAttrs.dataset;
```
Comment (matthew-white) [inline lib/data/dataset.js:79]:
```suggestion
    const datasetName = strippedAttrs.dataset?.trim();
```

File: lib/data/dataset.js [hunk 4]
```diff
@@ -61,21 +61,36 @@ const getDataset = (xml) => {
...
+    if (!validateDatasetName(datasetName))
       throw Problem.user.invalidEntityForm({ reason: 'Invalid dataset name.' });
-    return Option.of(datasetName.get().trim());
+
+    // Entity actions permitted by the form def
+    const actions = [];
+    const { create, update } = strippedAttrs;
+    if (create != null) actions.push('create');
+    if (update != null) actions.push('update');
+    if (actions.length === 0)
```
Comment (matthew-white) [inline lib/data/dataset.js:90]: Allow an empty array. As part of this, maybe update the Problem message.

File: lib/model/query/entities.js [hunk 5]
```diff
@@ -148,6 +148,14 @@ resolveConflict.audit = (entity, dataset) => (log) => log('entity.update.resolve
 /////////////////////////////////////////////////////////////////////////
 // Processing submission events to create and update entities
 
+const _getFormDefActions = async (oneFirst, datasetId, submissionDefId) => oneFirst(sql`
```
Comment (matthew-white) [inline lib/model/query/entities.js:151]: Pass this the `formDefId` from the `submissionDef`, then remove the `JOIN`.

File: test/unit/data/dataset.js [hunk 6]
```diff
@@ -53,7 +53,7 @@ describe('parsing dataset from entity block', () => {
                 <name/>
                 <age/>
                 <meta>
-                  <entities:entity dataset="foo">
+                  <entities:entity dataset="foo" create="">
```
Comment (matthew-white) [inline test/unit/data/dataset.js:56]: Revert.

File: test/unit/data/dataset.js [hunk 7]
```diff
@@ -162,40 +165,71 @@ describe('parsing dataset from entity block', () => {
...
-              </model>
+          <model entities:entities-version="2022.1.0">
+            <instance>
+              <data id="NoName">
+                <name/>
+                <age/>
+                <meta>
+                  <entities:entity create="">
+                    <label/>
+                  <entities:entity>
```
Comment (matthew-white) [inline test/unit/data/dataset.js:188]:
```suggestion
                  </entities:entity>
```

File: test/unit/data/dataset.js [hunk 8]
```diff
@@ -162,40 +165,71 @@ describe('parsing dataset from entity block', () => {
...
+      result.get().actions.should.eql(['create', 'update']);
+    });
+
+    it('should strip namespaces from action attributes', async () => {
+      const result = await getDataset(testData.forms.updateEntity
+        .replace('update=""', 'entities:create="" entities:update=""'));
+      result.get().actions.should.eql(['create', 'update']);
+    });
+
+    it('should reject for an entity form without an action', () => {
```
Comment (matthew-white) [inline test/unit/data/dataset.js:225]: Instead of rejecting, `actions` should be an empty array.

Let's also add an integration test that a submission is rejected if it attempts a create for a form that doesn't permit any entity actions.

**Review by ktuite — APPROVED** [review]

**Review by matthew-white — COMMENTED** [review]
Notes from interactive code review

File: lib/data/dataset.js (hunk 4 above)
Comment (matthew-white) [reply lib/data/dataset.js:90]: It actually turns out that the XForm spec requires at least one action attribute, so we'll continue to reject in that case.
